AI/Nullkiller2/Engine/PriorityEvaluator.cpp

```diff
@@ -0,0 +1,1696 @@
+{
+	auto objectsUnderThreat = aiNk->dangerHitMap->getOneTurnAccessibleObjects(enemy);
+	float objectValue = 0;
+
+	for(auto obj : objectsUnderThreat)
+	{
+		vstd::amax(objectValue, getStrategicalValue(obj));
+	}
+
+	/*
+	  1. If an enemy hero can attack nearby object, it's not useful to capture the object on our own.
+	  Killing the hero is almost as important (0.9) as capturing the object itself.
+
+	  2. The formula quickly approaches 1.0 as hero level increases,
+	  but higher level always means higher value and the minimal value for level 1 hero is 0.5
+	*/
+	return std::min(1.5f, objectValue * 0.9f + (1.5f - (1.5f / (1 + enemy->level))));
+}
+
+/// @return between 0-1.0f
+float RewardEvaluator::getNowResourceRequirementStrength(GameResID resType) const
+{
+	TResources requiredResources = aiNk->buildAnalyzer->getMissingResourcesNow();
+	TResources dailyIncome = aiNk->buildAnalyzer->getDailyIncome();
+
+	if(requiredResources[resType] == 0)
+		return 0;
+
+	if(dailyIncome[resType] == 0)
+		return 1.0f;
+
+	return 0.8f;
+}
+
+/// @return between 0-1.0f
+float RewardEvaluator::getTotalResourceRequirementStrength(GameResID resType) const
+{
+	TResources requiredResources = aiNk->buildAnalyzer->getMissingResourcesInTotal();
+	TResources dailyIncome = aiNk->buildAnalyzer->getDailyIncome();
+
+	if(requiredResources[resType] == 0)
+		return 0;
+
+	if(dailyIncome[resType] == 0)
+		return 1.0f;
+
+	return 0.8f;
+}
+
+uint64_t RewardEvaluator::townArmyGrowth(const CGTownInstance * town) const
+{
+	uint64_t result = 0;
+
+	for(auto creatureInfo : town->creatures)
+	{
+		if(creatureInfo.second.empty())
+			continue;
+
+		auto creature = creatureInfo.second.back().toCreature();
+		result += creature->getAIValue() * town->getGrowthInfo(creature->getLevel() - 1).totalGrowth();
+	}
+
+	return result;
+}
+
+float RewardEvaluator::getManaRecoveryArmyReward(const CGHeroInstance * hero) const
+{
+	return aiNk->heroManager->getMagicStrength(hero) * 10000 * (1.0f - std::sqrt(static_cast<float>(hero->mana) / hero->manaLimit()));
+}
+
+/// @return between 0-1.0f
+float RewardEvaluator::getCombinedResourceRequirementStrength(const TResources & res) const
+{
+	float sum = 0.0f;
+
+	for(TResources::nziterator it(res); it.valid(); it++)
+	{
+		auto calculation = 0.5f * getNowResourceRequirementStrength(it->resType)
+			+ 0.5f * getTotalResourceRequirementStrength(it->resType);
+
+		// Even not required resources should be valuable because they shouldn't be left for the enemies to collect
+		sum += std::min(MINIMUM_STRATEGICAL_VALUE_NON_TOWN, calculation);
+	}
+
+	return sum;
+}
+
+float RewardEvaluator::getStrategicalValue(const CGObjectInstance * target, const CGHeroInstance * hero) const
+{
+	if(!target)
+		return 0;
+
+	switch(target->ID)
+	{
+	case Obj::MINE:
+	{
+		auto mine = dynamic_cast<const CGMine *>(target);
+		TResources res;
+		res[mine->producedResource] = mine->producedQuantity;
+
+		// Mines should have higher priority than resources
```

Comment:
You're going to have a looong talk with competitive MP players :)

(in competitive MP maps are often very resource-rich, so mines are very low in priority compared to resource piles)